### ISSUE TITLE
TypeError: Cannot read property 'opened' of undefined

### ISSUE THREAD

**Issue by eamello** [phanxgames/phanx-mysql#2]
Intermittently, this error shows up and crashes my express server in Node.JS. 

TypeError: Cannot read property 'opened' of undefined
    at Timeout._onTimeout (E:\Repo CBGA\cbga-backend\node_modules\phanx-mysql\PhanxMysql.ts:118:28)
    at listOnTimeout (internal/timers.js:554:17)
    at processTimers (internal/timers.js:497:7)
[ERROR] 17:57:37 TypeError: Cannot read property 'opened' of undefined

Already considering to ditch this MySQL database wrapper and move to another one, where the author can respond to issues, etc.

**Comment by phanxgames** [comment]
I apologize for not seeing your previous issue last month. I completely understand if you wish to move on to another wrapper, but if you are willing to help me resolve this issue, I would like to make this package the best it can be.

So if you could kindly let me know your environment details:
NodeJS version, express version, any other information you are willing to share.

Perhaps a code example you are using (if you are able to share).

I use this wrapper for all of my projects, and I have never seen this issue. But I am more than happy to look into it for you.

Thank you for your interest of the library (even if former) and any insight you can share to resolve this.

**Comment by eamello** [comment]
Hi Henry, 
Thanks for getting back. 
I like your wrapper and I am not planning to change to another one. 
However this problem has been happening Intermittently and it crashes my express server. It's something to do with "autoCloseMinutes" - it's trying to close a connection where the db object is undefined.

NodeJS version: v14.15.0
Express: 4.17.1

For every single call, I open and close the db like the code below (The BaseService class can be seen on the other issue about query_format, I pasted it in there).

export class TechLogService extends BaseService { 

	constructor() {
		super();
	}

	public async create(objEntity: TechLog, decodedUser: DecodedUser): Promise<number> {

		let new_id = 0;

		if(objEntity){

			await this.db.start();
			new_id = await this.db.insertAndRun(TechLog.TABLE_NAME, this.getKeyValuePairToInsert(objEntity, decodedUser.payload.user_id, TechLog.PRIVATE_KEY));
			await this.db.end();
		}

		return new_id;
	}

	public async update(objEntity: TechLog, primary_key_id: number, decodedUser: DecodedUser): Promise<boolean> {

		await this.db.start();
		await this.db.updateAndRun(TechLog.TABLE_NAME, TechLog.PRIVATE_KEY+"=?", [primary_key_id], this.getKeyValuePairToUpdate(objEntity, decodedUser.payload.user_id, TechLog.PRIVATE_KEY));
		await this.db.end();
		return true;
	}

**Comment by phanxgames** [comment]
I believe this is some kind of race condition between the auto-closer and the reuse of connection instances. In my production environment I do not reuse the connection instance, but instead call PhanxMysql.createAndStart() each time I wish to open a connection. 

So instead of writing:
```
await this.db.start();
...
await this.db.end();
```
I would instead write:
```
let db = await PhanxMysql.createAndStart();
...
db.end();
```
This does not cause any issues, and minimal performance hit, because the connection pooling internally reuses connections.

However, I agree that your style and method is valid and should not cause this issue. 

I have made an update to the code, which I will publish shortly, that will solve the null exception, which will solve the crashing. But the underlying issue is not resolved.

I would suggest either doing what I suggest above in this post, or disable the auto-close functionality by making sure you pass 0 minutes to the method and the config. I noticed you have it in your config and you are passing it directly as well. Make sure both are 0 to disable the auto-close functionality.  The auto-close is not required if you always call .end for every .start.
